src/Common/Core/Impl/Threading/AsyncReaderWriterLock.Queue.cs

```diff
@@ -134,7 +134,7 @@ private class Queue {
             public LockSource[] Remove(LockSource item) {
                 lock (this) {
                     var next = item.Next;
-                    Interlocked.CompareExchange(ref _head, next, item);
+                    var oldHead = Interlocked.CompareExchange(ref _head, next, item);
```

Comment:
I mean the operation looks like concurrent double-linked list. Maybe just make a class?